07_errors/hsy3418/mapStore.go

```diff
@@ -0,0 +1,53 @@
+package main
+
+// MapStore is a implementation of storer for the storage of puppies
+type MapStore struct {
+	puppies map[int32]Puppy
+}
+
+func NewMapStore() *MapStore {
+	return &MapStore{
+		puppies: map[int32]Puppy{},
+	}
+}
+
+// CreatePuppy adds a nuw puppy to the puppies store
+func (m *MapStore) CreatePuppy(puppy Puppy) error {
+	if puppy.Value < 0 {
+		return ErrorEf(ErrInvalidInput, "The puppy value is invalidate")
+	}
+	if _, exists := m.puppies[puppy.ID]; !exists {
```

Comment:
A couple of things here, I mentioned in the previous pr that normal convention for golang is **if error -> fail early**, and that I was a little more relaxed on it when there is only a single error constraint. Now that this has multiple I think this should be flipped so that the final return is the puppy and the error duplicate being returned if the puppy exists.

Given that, I also think that this should not be the case where the ID is passed in the puppy and validated on creation. I think an ID should be assigned and returned back to the user in the response. This was raised in your previous PR, so when you update it there can you please bring it through to this one as well.